### PR TITLE
feat: implement encoding to fit

### PR DESCRIPTION
**Pull request by muktihari**
- implement encoding into fit file format
- create building block for gpx and tcx, implement later
- create UI element for Tools and wire it to WASM encoding
- update manufacturers.json
- fit minor bug on Session's AvgPace calculation
- fix minor bug on graphs

Tools preview:
image

image

image

image

### CODE REVIEW

**Comment by raditzlawliet** [comment]
1. 2nd and 3rd tools need indicator that this tools is disabled for single file. The flow works, but in chrome the disabled selection looks same can cause confusion. 
image
sample when using 2 files image

2. When Conceal GPS is selected, the button is disabled. if it required user to select the distance. please add some indicator like warning or required that user need to select distance of conceal GPS. OR just enable it
image

3. (Optional) Maybe distance of Conceal GPS can show some marker in map like in strava
4. There's an error when select trim track point with single file using tools edit
image
image

**Comment by muktihari** [comment]
1. This is tricky since I use default html select  styled using bootstrap, look good on mac chrome, need to find other select component later. Do you have any suggestion?
image
2. Let's just enable it for now since it seems to cause confusion.
3. This is on roadmap, will update it on next PR. Need to find a way to show map in mobile as well since the map location in mobile is too far away.
4. How can I reproduce this? Please give step by step reproduce procedure since I try in both chrome and firefox seem ok.

**Comment by muktihari** [comment]
I have update regarding the disabled button ya.

btw, actually on mac, the select look better than in firefox, lol
chrome:
image

firefox:
image

**Comment by raditzlawliet** [comment]
seems no 4 not reliable how to produce, i retried again now with same procedure and no issue. you can ignore no 4.

No 1, about disabled option in selection. Not sure, the UI Select on windows chrome still same 
image
that from edge
image
that from chrome

the disabled option i mean, when i try to select 1 file, the option tools for combine and split can't be selected. so i assume it should be disabled right?

and, we can try to put this to another issue relate UI/UX instead.

File: src/components/TheTools.vue [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+  },
+  data() {
+    return {
+      manufacturerMap: shallowRef(new Map<number, Manufacturer>()),
+      sportMap: shallowRef(new Map<string, Sport>()),
+      EncodeMode: EncodeMode,
+      encodeMode: EncodeMode.Unknown,
+      selectedManufacturerProduct: null as unknown as any[],
+      manufacturerName: '',
+      manufacturerId: null as unknown as number | null,
```
Comment (raditzlawliet) [inline src/components/TheTools.vue:358]: is it really unknown?

File: src/spec/activity-service.ts [hunk 2]
```diff
@@ -20,22 +20,72 @@ export class DecodeResult {
...
 export class EncodeResult {
   err: string | null = null
+  deserializeInputTook: number
   encodeTook: number
   serializationTook: number
   totalElapsed: number
   fileName: string
   fileType: string
-  fileBytes: Uint8Array
+  filesBytes: Uint8Array
```
Comment (raditzlawliet) [inline src/spec/activity-service.ts:29]: just wondering, is it bytes of multiple files?

File: src/wasm/activity-service/activity/fit/session.go [hunk 3]
```diff
@@ -155,3 +158,131 @@ func NewSession(mesg proto.Message) *activity.Session {
...
+		field.Value = scaleoffset.DiscardAny(ses.TotalElapsedTime, field.Type.BaseType(), field.Scale, field.Offset)
+		mesg.Fields = append(mesg.Fields, field)
+
+		totalTimerTimeField := factory.CreateField(mesgnum.Lap, fieldnum.SessionTotalElapsedTime)
+		totalTimerTimeField.Value = field.Value
+		mesg.Fields = append(mesg.Fields, totalTimerTimeField)
+	}
+	if ses.TotalDistance != 0 {
+		field := factory.CreateField(mesgnum.Session, fieldnum.SessionTotalDistance)
+		field.Value = scaleoffset.DiscardAny(ses.AvgAltitude, field.Type.BaseType(), field.Scale, field.Offset)
```
Comment (raditzlawliet) [inline src/wasm/activity-service/activity/fit/session.go:200]: how is TotalDistance relate with AvgAltitude?

File: src/wasm/activity-service/kit/kit.go [hunk 4]
```diff
@@ -42,3 +42,38 @@ func FormatTitle(s string) string {
...
+		return b
+	}
+	if b != nil {
+		c := T((float64(*a) + float64(*b)) / 2)
+		return &c
+	}
+	return a
+}
+
+// Avg returns max of two non-nil values. Otherwise, return any non-nil value if possible.
```
Comment (raditzlawliet) [inline src/wasm/activity-service/kit/kit.go:70]: nitpick, typo comment

**Review by raditzlawliet — COMMENTED** [review]
I have Some question

**Comment by raditzlawliet** [comment]
Is Split Activity per Session need 2 files? or can be use even single file?

**Comment by muktihari** [comment]
Select: What I get from the internet, `select` styling is OS-based not a browser-based style, so it doesn't matter what browser we use, but what OS we use. I will update it using `vue-select` for more reliable styling and enable search as well.

File: src/components/TheTools.vue (hunk 1 above)
Comment (muktihari) [reply src/components/TheTools.vue:358]: it's just to make manufacturerId type is `number | null`, we we can't do it straightforwardly in data(), need to convert it to unknown first and cast it back

File: src/spec/activity-service.ts (hunk 2 above)
Comment (muktihari) [reply src/spec/activity-service.ts:29]: yes its array of bytes ([][]byte{}), but it looks like I haven't update the typescript side, nice catch 👍

File: src/wasm/activity-service/activity/fit/session.go (hunk 3 above)
Comment (muktihari) [reply src/wasm/activity-service/activity/fit/session.go:200]: ah you found a bug, it should be ses.TotalDistance

**Comment by muktihari** [comment]
Update on dropdown selection, now with more consistent style across browser and OS.
image

**Comment by muktihari** [comment]
> Is Split Activity per Session need 2 files? or can be use even single file?

If a single activity contains more than one session, then the split mode is enabled.

**Comment by muktihari** [comment]
Update help text so user will know why it is disabled at first glance

image

**Comment by raditzlawliet** [comment]
image
if product_id field is mandatory, can you put info or something that it is required?

**Comment by raditzlawliet** [comment]
> image if product_id field is mandatory, can you put info or something that it is required?

like red asterisk? or something red message it required

**Comment by muktihari** [comment]
actually all fields are required, but in case user confused why can't proceed after selecting device, I added the asterisk as requested.

File: src/spec/activity-service.ts (hunk 2 above)
Comment (muktihari) [reply src/spec/activity-service.ts]: This has been updated too ya, I think this PR is complete, do you have any more concern?

**Pull request merged**